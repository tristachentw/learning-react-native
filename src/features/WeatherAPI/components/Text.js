import styled, { withTheme } from 'styled-components'

const CustomText = styled.Text`
  font-size: ${props => {
    const { size, h1, h2, h3 } = props
    const { fontSize, h1FontSize, h2FontSize, h3FontSize } = props.theme
    if (size) {
      return size
    } else if (h1) {
      return h1FontSize
    } else if (h2) {
      return h2FontSize
    } else if (h3) {
      return h3FontSize
    }
    return fontSize
  }};
  font-weight: ${props => {
    const { weight, h1, h2, h3 } = props
    return weight || (h1 || h2 || h3) ? 'bold' : 'normal'
  }};
  color: ${props => props.color || props.theme.primaryColor};
`

CustomText.defaultProps = {
  theme: {
    h1FontSize: 32,
    h2FontSize: 24,
    h3FontSize: 18,
    fontSize: 14,
    primaryColor: '#000'
  }
}

export default withTheme(CustomText)
