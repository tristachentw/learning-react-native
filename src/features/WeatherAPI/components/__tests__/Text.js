import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { ThemeProvider } from 'styled-components/native'

import Text from '../Text'

const THEME = {
  h1FontSize: 40,
  h2FontSize: 30,
  h3FontSize: 20,
  fontSize: 10,
  primaryColor: 'black'
}

const renderWithTheme = (children, theme) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}

describe('Custom Text Component', () => {
  it('should render without throwing an error', function () {
    const component = shallow(<Text />)
    expect(toJson(component)).toMatchSnapshot()
  })

  it('should render with theme without throwing an error', function () {
    const component = renderWithTheme(<Text />, THEME)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should have heading styles provided by theme', () => {
    const component = renderWithTheme(<Text h1 />, THEME)
    const view = component.root.findByType('Text')
    expect(view.props.style).toEqual([{
      fontSize: 40,
      fontWeight: 'bold',
      color: 'black'
    }])
  })

  it('should have text as children', () => {
    const component = renderWithTheme(<Text>Children Text</Text>, THEME)
    const view = component.root.findByType('Text')
    expect(view.props.children).toBe('Children Text')
  })
})
