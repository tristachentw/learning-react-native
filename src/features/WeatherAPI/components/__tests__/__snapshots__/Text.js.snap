// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Text Component should render without throwing an error 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "black",
        "fontSize": 10,
        "fontWeight": "normal",
      },
    ]
  }
  theme={
    Object {
      "fontSize": 14,
      "h1FontSize": 32,
      "h2FontSize": 24,
      "h3FontSize": 18,
      "primaryColor": "#000",
    }
  }
/>
`;
